kernel/gdt.c

```diff
@@ -41,15 +44,113 @@ static void gdt_set_gate(int num, uint32_t base, uint32_t limit, uint8_t access,
 	gdtp.base = (uint32_t)gdt;
-	gdt_set_gate(0,	0,	0,	0,	0);	// NULL descriptor
+	gdt_set_gate(0,	0,	0x0000,	0,	0);	// NULL descriptor
 	gdt_set_gate(1,	0,	0xFFFF,	0x9A,	0xCF);	// Kernel Mode Code Segment
 	gdt_set_gate(2,	0,	0xFFFF,	0x92,	0xCF);	// Kernel Mode Data Segment
 	gdt_set_gate(3,	0,	0xFFFF,	0x96,	0xCF);	// Kernel Mode Stack Segment
 	gdt_set_gate(4,	0,	0xFFFF,	0xFA,	0xCF);	// User Mode Code Segment
 	gdt_set_gate(5,	0,	0xFFFF,	0xF2,	0xCF);	// User Mode Data Segment
-	gdt_set_gate(6,	0,	0xFFFF,	0x96,	0xCF);	// User Mode Stack Segment
+	gdt_set_gate(6,	0,	0xFFFF,	0xF6,	0xCF);	// User Mode Stack Segment
 	gdt_flush(&gdtp);
 }
 
+/* Print all the info from the gdt basic info and all of the entries
+ *
+ */
+void print_gdt()
+{
+	VGA_setforegroundcolor(VGA_COLOR_GREEN);
+	kprintf("GDT ADDRESS = 0x00000%x\n", gdt);
+	VGA_setforegroundcolor(VGA_COLOR_WHITE);
+	kprintf("GDT ENTRIES : \n");
+	kprintf("GDT Base = 0x%x\n", gdtp.base);
+	kprintf("GDT Limit = 0x%x\n", gdtp.limit);
+	uint8_t i = 1;
+	VGA_setforegroundcolor(VGA_COLOR_LIGHT_BROWN);
+	kprintf("BASE LOW | BASE MID | BASE HIGH | LIMIT LOW | GRAN | ACC\n");
+	VGA_setforegroundcolor(VGA_COLOR_WHITE);
+	kprintf("0x%x        ", gdt[0].base_low);
+	kprintf("0x%x        ", gdt[0].base_middle);
+	kprintf("0x%x        ", gdt[0].base_high);
+	kprintf("0x%x           ", gdt[0].limit_low);
+	kprintf("0x%x         ", gdt[0].granularity);
+	kprintf("0x%x\n", gdt[0].access);
+	while (i < 7)
+	{
+		kprintf("0x%x        ", gdt[i].base_low);
+		kprintf("0x%x        ", gdt[i].base_middle);
+		kprintf("0x%x        ", gdt[i].base_high);
+		kprintf("0x%x        ", gdt[i].limit_low);
+		kprintf("0x%x        ", gdt[i].granularity);
+		kprintf("0x%x\n", gdt[i].access);
+		i++;
+	}
+}
+
+/* Print the stack from ebp to limit
+ *
+ * No args at the moment.
+ * @ebp is the base pointer from the stack
+ * @limit is the mumber of byte printed (one address == 16 bytes) and it's
+ * hardcoded. In the future we can set it as parameter.
+ *
+ */
+void print_stack()
+{
+	uint32_t ebp;
+	asm inline volatile ("mov %%ebp, %0" :: "r"(ebp));
+	uint32_t limit = 4096;
+	uint32_t i = 0;
+	uint8_t line = 0;
+	struct kbd_event evt;
+
+	while (i < limit)
+	{
+		if (line == 25)
+		{
+			while (!KBD_poll());
+			KBD_getevent(&evt);
+			if (evt.type != KEY_PRESSED)
+				continue;
+			line = 0;
+		}
+		VGA_setforegroundcolor(VGA_COLOR_WHITE);
+		kprintf("0x%x : ", ebp);
+		uint32_t next = ebp + 16;
+		uint32_t tmp = ebp;
+
+		while (tmp < next)
+		{
+			if (*(char*)tmp > 32)
+			{
+				VGA_setforegroundcolor(VGA_COLOR_BLUE);
+				kprintf("%x ", *(char*)tmp);
+				VGA_setforegroundcolor(VGA_COLOR_WHITE);
+			}
+			else
+				kprintf("00 ");
+			tmp++;
+		}
+		next = ebp + 16;
+		tmp = ebp;
+		while (tmp < next)
+		{
+			if (*(char*)tmp > 32)
+			{
+				VGA_setforegroundcolor(VGA_COLOR_BLUE);
+				kprintf("%c", *(char *)tmp);
+				VGA_setforegroundcolor(VGA_DFL_COLOR);
+			}
+			else
+				kprintf(".");
+			tmp++;
+		}
```

Comment:
- This block is used twice with a minor difference. Consider extracting it to a static function, taking a string as parameter to print in case of a 0 for example. (You can also take the format string to give to `kprintf` as parameter).

- next and tmp are assigned twice but they doesn't change between the two assignments.